### PR TITLE
Implemented pagination feature in erlcloud_ec2:describe_instances

### PR DESCRIPTION
**Pull request by SergeyOvchinnik**
Problem: 
Function describe_instances in erlcloud_ec2 did not support pagination feature on library level

Solution:
Implemented pagination feature for erlcloud_ec2:describe_instances function

Reviewers:
@ddosia @motobob

### CODE REVIEW

File: src/erlcloud_ec2.erl [hunk 1]
```diff
@@ -169,8 +169,11 @@
          describe_vpn_gateways/0, describe_vpn_gateways/1, describe_vpn_gateways/2, describe_vpn_gateways/3,
          describe_vpn_connections/0, describe_vpn_connections/1, describe_vpn_connections/2, describe_vpn_connections/3,
         
-        %% Customer gateways
-         describe_customer_gateways/0, describe_customer_gateways/1, describe_customer_gateways/2, describe_customer_gateways/3
+         %% Customer gateways
+         describe_customer_gateways/0, describe_customer_gateways/1, describe_customer_gateways/2, describe_customer_gateways/3,
+
+         %% Testing
+         describe_instances_test/0
```
Comment (motobob) [inline src/erlcloud_ec2.erl]: put it in tests and not into code. see test/erlcloud_ec2_tests.erl

File: src/erlcloud_ec2.erl [hunk 2]
```diff
@@ -1347,38 +1350,101 @@ attribute_xpath(block_device_mapping, AttributeName) ->
...
-    describe_instances(InstanceIDs, [], default_config()).
-
--spec(describe_instances/2 :: ([string()], filter_list() | aws_config()) -> proplist()).
+    when is_record(Config, aws_config) ->
+    describe_instances([], [], Config);
+describe_instances(InstanceIDs)
+    when is_list(InstanceIDs) ->
+    describe_instances(InstanceIDs, [], default_config());
+describe_instances(MaxResults)
+    when is_integer(MaxResults) andalso MaxResults >= 5 andalso MaxResults =< 1000 orelse MaxResults =:= undefined ->
```
Comment (motobob) [inline src/erlcloud_ec2.erl]: let not allow 
```erlang
orelse MaxResults =:= undefined 
```

File: src/erlcloud_ec2.erl [hunk 3]
```diff
@@ -1347,38 +1350,101 @@ attribute_xpath(block_device_mapping, AttributeName) ->
 attribute_xpath(_, AttributeName) ->
     "/DescribeInstanceAttributeResponse/" ++ AttributeName ++ "/value".
 
+-type ec2_token() :: string() | undefined.
+-type ec2_instances_ids() :: [string()].
+-type ec2_max_result() :: integer() | undefined.
```
Comment (motobob) [inline src/erlcloud_ec2.erl]: make it `5..1000`?

File: src/erlcloud_httpc.erl [hunk 4]
```diff
@@ -43,10 +43,8 @@ request(URL, Method, Hdrs, Body, Timeout,
     when is_function(F, 6) ->
     F(URL, Method, Hdrs, Body, Timeout, Config).
 
-request_lhttpc(URL, Method, Hdrs, Body, Timeout, #aws_config{lhttpc_pool = undefined}) ->
-    lhttpc:request(URL, Method, Hdrs, Body, Timeout, []);
-request_lhttpc(URL, Method, Hdrs, Body, Timeout, #aws_config{lhttpc_pool = Pool}) ->
-    lhttpc:request(URL, Method, Hdrs, Body, Timeout, [{pool, Pool}]).
+request_lhttpc(URL, Method, Hdrs, Body, Timeout, _Config) ->
```
Comment (motobob) [inline src/erlcloud_httpc.erl]: breaking unnecessary change

File: src/erlcloud_ec2.erl [hunk 5]
```diff
@@ -1347,38 +1350,101 @@ attribute_xpath(block_device_mapping, AttributeName) ->
...
+    when is_integer(MaxResults) andalso MaxResults >= 5 andalso MaxResults =< 1000 orelse MaxResults =:= undefined ->
+    describe_instances(MaxResults, undefined, default_config()).
+
+-spec describe_instances(ec2_instances_ids(), filter_list()) -> ec2_describe_return();
+                        (ec2_instances_ids(), aws_config()) -> ec2_describe_return();
+                        (ec2_max_result(), aws_config()) -> ec2_describe_return();
+                        (ec2_max_result(), ec2_token()) -> ec2_describe_return().
+describe_instances(InstanceIDs, Filter)
+    when is_list(InstanceIDs), is_list(Filter) ->
+    describe_instances(InstanceIDs, Filter, default_config());
```
Comment (motobob) [inline src/erlcloud_ec2.erl]: a filtered request can still be very big and requires pagination. this is not implemented

File: src/erlcloud_ec2.erl [hunk 6]
```diff
@@ -1347,38 +1350,101 @@ attribute_xpath(block_device_mapping, AttributeName) ->
...
 
+%% TODO: move to test/erlcloud_ec2_tests
+
+describe_instances_test() ->
+    {ok, AllInstances} = describe_instances(),
+    {ok, PagedInstances} = describe_instances_all(),
+    ListsSame = lists_are_the_same(AllInstances, PagedInstances),
+    {'Original',length(AllInstances),'Paged', length(PagedInstances), ListsSame}.
+
+describe_instances_all() ->
```
Comment (motobob) [inline src/erlcloud_ec2.erl]: this should be part of normal API.

File: src/erlcloud_ec2.erl [hunk 7]
```diff
@@ -1347,38 +1350,101 @@ attribute_xpath(block_device_mapping, AttributeName) ->
...
+  {ok, describe_instances_all([], undefined)}.
+describe_instances_all(List, NextToken) ->
+    case describe_instances(40, NextToken) of
+      {ok, NextPage,  undefined} ->
+          lists:append(List, NextPage);
+      {ok, NextPage, NT} ->
+          describe_instances_all(lists:append(List, NextPage), NT)
+    end.
+
+lists_are_the_same(List1, List2) ->
```
Comment (motobob) [inline src/erlcloud_ec2.erl]: this ignores the order of elements in the list.
I'm not sure  [11,22] and [22,11] can be called the same;)

File: src/erlcloud_ec2.erl [hunk 8]
```diff
@@ -1347,38 +1350,101 @@ attribute_xpath(block_device_mapping, AttributeName) ->
...
+    length(List1) == length(List2) andalso
+    is_sublist(List1, List2) andalso
```
Comment (motobob) [inline src/erlcloud_ec2.erl]: this is extremely non performant. 
hint: L1- L2 =:=0 ;)

File: src/erlcloud_ec2.erl (hunk 7 above)
Comment (motobob) [inline src/erlcloud_ec2.erl]: also why not just 
`L1 == L2` ?

File: src/erlcloud_ec2.erl (hunk 6 above)
Comment (motobob) [inline src/erlcloud_ec2.erl]: meaning describe_instances() and describer_instances_all() are different. ;)

File: src/erlcloud_ec2.erl [hunk 9]
```diff
@@ -1347,38 +1350,101 @@ attribute_xpath(block_device_mapping, AttributeName) ->
...
+
+describe_instances_test() ->
+    {ok, AllInstances} = describe_instances(),
+    {ok, PagedInstances} = describe_instances_all(),
+    ListsSame = lists_are_the_same(AllInstances, PagedInstances),
+    {'Original',length(AllInstances),'Paged', length(PagedInstances), ListsSame}.
+
+describe_instances_all() ->
+  {ok, describe_instances_all([], undefined)}.
+describe_instances_all(List, NextToken) ->
```
Comment (motobob) [inline src/erlcloud_ec2.erl]: please notice `erlcloud_util:query_all()`

File: src/erlcloud_ec2.erl (hunk 2 above)
Comment (SergeyOvchinnik) [inline src/erlcloud_ec2.erl]: Agree, removed it

File: src/erlcloud_ec2.erl (hunk 1 above)
Comment (SergeyOvchinnik) [inline src/erlcloud_ec2.erl]: Will do

File: src/erlcloud_ec2.erl (hunk 5 above)
Comment (SergeyOvchinnik) [inline src/erlcloud_ec2.erl]: MaxResults and InstanceIDs cannot be specified in the same call. And in the original version of describe_instances the Filter was specified only when InstanceIDs is specified. Would you like me to still implement the Filter and MaxResults combination?

File: src/erlcloud_ec2.erl (hunk 7 above)
Comment (SergeyOvchinnik) [inline src/erlcloud_ec2.erl]: Because pagination sometimes breaks the order

File: src/erlcloud_ec2.erl (hunk 3 above)
Comment (SergeyOvchinnik) [inline src/erlcloud_ec2.erl]: I was actually intending to use this type for all describe functions and only specify borders in guards since borders are different for different functions

File: src/erlcloud_ec2.erl (hunk 6 above)
Comment (SergeyOvchinnik) [inline src/erlcloud_ec2.erl]: So far I only used the _all version to test pagination. Would you like me to implement the _all function as just a general function which returns everything but makes lots of paginated calls in the process instead of one?

File: src/erlcloud_ec2.erl (hunk 8 above)
Comment (SergeyOvchinnik) [inline src/erlcloud_ec2.erl]: Ok, will try that

File: src/erlcloud_ec2.erl (hunk 9 above)
Comment (SergeyOvchinnik) [inline src/erlcloud_ec2.erl]: Honestly, I don't see much point in <code>erlcloud_util:query_all()</code>. It attaches some unneeded params to the call like Marker and MaxItems instead of MaxResults. So unless I rewrite it, it will just receive UnknownParameter error.

File: src/erlcloud_httpc.erl (hunk 4 above)
Comment (SergeyOvchinnik) [inline src/erlcloud_httpc.erl]: true, forgot to remove it in this commit

File: src/erlcloud_ec2.erl (hunk 5 above)
Comment (motobob) [inline src/erlcloud_ec2.erl]: yes. that's my point under "filtered request"

File: src/erlcloud_ec2.erl (hunk 6 above)
Comment (motobob) [inline src/erlcloud_ec2.erl]: yes

File: src/erlcloud_ec2.erl (hunk 9 above)
Comment (motobob) [inline src/erlcloud_ec2.erl]: `erlcloud_util:query_all_token()` is also similar. 
I guess my point is that you will find useful having a generalised `query_all` which works with `MaxResults`. 
So i would suggest to probably spend some time on this. probably worth separate PR but IMO that will be noticeably useful even within just ec2 module

File: src/erlcloud_ec2.erl (hunk 7 above)
Comment (motobob) [inline src/erlcloud_ec2.erl]: not sure I get it. how come?

File: src/erlcloud_ec2.erl (hunk 5 above)
Comment (SergeyOvchinnik) [inline src/erlcloud_ec2.erl]: Ok, will implement in next commit

File: src/erlcloud_ec2.erl (hunk 6 above)
Comment (SergeyOvchinnik) [inline src/erlcloud_ec2.erl]: I'm on it

File: src/erlcloud_ec2.erl (hunk 9 above)
Comment (SergeyOvchinnik) [inline src/erlcloud_ec2.erl]: Ok, I'll see what can be done there

File: src/erlcloud_ec2.erl [hunk 10]
```diff
@@ -1347,38 +1350,101 @@ attribute_xpath(block_device_mapping, AttributeName) ->
...
+    when is_record(Config, aws_config) ->
+    describe_instances([], [], Config);
+describe_instances(InstanceIDs)
+    when is_list(InstanceIDs) ->
+    describe_instances(InstanceIDs, [], default_config());
+describe_instances(MaxResults)
+    when is_integer(MaxResults) andalso MaxResults >= 5 andalso MaxResults =< 1000 orelse MaxResults =:= undefined ->
+    describe_instances(MaxResults, undefined, default_config()).
+
+-spec describe_instances(ec2_instances_ids(), filter_list()) -> ec2_describe_return();
```
Comment (ddosia) [inline src/erlcloud_ec2.erl]: Am I right that call without `ec2_max_result()` as an argument will never return `{ok, proplist(), ec2_token()}`?
If so then spec is too wide. I should be `{ok, proplist()}` in one clause and `{ok, proplist(), ec2_token()}` in another.

File: src/erlcloud_ec2.erl [hunk 11]
```diff
@@ -1347,38 +1350,101 @@ attribute_xpath(block_device_mapping, AttributeName) ->
 attribute_xpath(_, AttributeName) ->
     "/DescribeInstanceAttributeResponse/" ++ AttributeName ++ "/value".
 
+-type ec2_token() :: string() | undefined.
+-type ec2_instances_ids() :: [string()].
+-type ec2_max_result() :: integer() | undefined.
+-type ec2_describe_return() :: {ok, proplist()} | {ok, proplist(), ec2_token()}.
```
Comment (SergeyOvchinnik) [inline src/erlcloud_ec2.erl]: Yes, added that

File: src/erlcloud_ec2.erl (hunk 10 above)
Comment (SergeyOvchinnik) [inline src/erlcloud_ec2.erl]: Good point, I'll fix it now...

**Comment by zdaniel86** [comment]
As erlcloud user I would like the callback syntax:
```Erlang
my_function() ->
   ok = erlcloud_ec2:describe_instances_all(fun do_smth_with_instance/1).
```
I just want that erlcoud calls me with each instance instead.

**Comment by motobob** [comment]
@zdaniel86 such API is not in the library in any module so it's unclear why would we introduce this here. 
i guess same need then for 
```erlang
erlcloud_ec2:describe_instances(fun foo_bar_callback/1),
```
Also, what is the use case for such an API? 
I would suggest opening a separate track for that if really need this.

**Comment by SergeyOvchinnik** [comment]
@zdaniel86 @motobob I mean, it doesn't seem like a big task at all. If we consider it being a useful feature, I can introduce it to each describe function in the module once I finish pagination.

**Comment by motobob** [comment]
it can be super simple but i still do see value in such so let's finish current task first and then switch to adding new APIs

**Comment by SergeyOvchinnik** [comment]
Made a new commit with some changes, currently working on moving the test to erlcloud_ec2_tests
